servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java

```diff
@@ -237,15 +235,28 @@ private void onError0(Throwable throwable, boolean cancelUpstream) {
             }
         }
 
+        private void onErrorHoldingLock(Throwable cause) {
+            try {
+                doCancel(true);
+            } finally {
+                sendToTarget(error(cause));
+            }
+        }
+
         private void tryEmitItem(final Object item) {
             if (tryAcquireLock(emittingUpdater, this)) { // fast path. no concurrency, avoid the queue and emit.
                 try {
                     sendToTarget(item);
+                    // Call doDrainPostProcessing inside the lock so if an exception is thrown we can propagate an
+                    // error terminal downstream and ignore future signals.
                     doDrainPostProcessing(1);
-                } finally {
-                    if (!releaseLock(emittingUpdater, this)) {
-                        drainPending();
-                    }
+                } catch (Throwable cause) {
+                    onErrorHoldingLock(cause);
+                    return; // Poison emittingUpdater. We prematurely terminated, other signals should be ignored.
+                }
+                // Release lock after we handle errors, because error handling needs to poison the lock.
+                if (!releaseLock(emittingUpdater, this)) {
```

Comment:
Here is a risk that smth inside `onErrorHoldingLock` may throw and then lock won't be released

Reply:
that is intentional. when `onErrorHoldingLock` is called the lock is "poisoned" (aka not unlocked) to prevent future signal delivery.

Reply:
I see, thanks for clarifying!